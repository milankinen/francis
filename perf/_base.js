const memwatch = require("@airbnb/node-memwatch")
const Benchmark = require("benchmark")
const bytes = require("pretty-bytes")
const { format } = require("util")
const { Francis, Bacon, Kefir } = require("./_libs")

const MEMORY = !!process.env.SHOW_MEMORY_USAGE
let heapStats = []
let gcTime = 0

const withFrancis = f => (...args) => f(Francis, ...args)
const withBacon = f => (...args) => f(Bacon, ...args)
const withKefir = f => (...args) => f(Kefir, ...args)

const resetHeapStats = () => {
  gcTime = 0
  heapStats = []
}

const getAvgMemoryUsage = stats =>
  stats.length ? bytes(Math.floor(stats.reduce((a, s) => a + s, 0) / stats.length)) : "unknown"
const getMaxMemoryUsage = stats =>
  stats.length ? bytes(Math.floor(stats.reduce((a, s) => Math.max(a, s), 0))) : "unknown"

if (MEMORY) {
  memwatch.on("stats", stats => {
    heapStats.push(stats.used_heap_size)
    gcTime += stats.gc_time / 1000000.0
  })
}

const testCases = (label, fn, combinations = []) => {
  const accepted = (testFn, args) => !testFn.accept || testFn.accept(...args)
  const cases = (name, testFn) =>
    combinations
      .map(
        args =>
          accepted(testFn, args) && {
            description:
              `${name} ${format(label + "|||", ...args).split("|||")[0]}` +
              (testFn.note ? ` (${testFn.note})` : ""),
            defer: true,
            fn: d => {
              const done = () => d.resolve()
              testFn(...args, done)
            },
          },
      )
      .filter(x => x)
  return [
    ...(process.env.NO_FRANCIS ? [] : cases("francis", withFrancis(fn))),
    ...(process.env.NO_BACON ? [] : cases("bacon  ", withBacon(fn))),
    ...(process.env.NO_KEFIR || !fn.kefir ? [] : cases("kefir  ", withKefir(fn.kefir))),
  ]
}

const runTests = (suiteName, testCases) => {
  return new Promise(resolve => {
    console.log(`Running suite ${suiteName}...`)
    console.log("")
    const maxLen = testCases.reduce((m, t) => Math.max(t.description.length, m), 0)
    resetHeapStats()
    testCases
      .reduce(
        (suite, test) => suite.add(test.description.padEnd(maxLen + 1), test),
        new Benchmark.Suite(),
      )
      .on("complete", function() {
        if (MEMORY) {
          const stats = [...heapStats]
          console.log("Memory statistics:")
          console.log(" > Avg mem:", getAvgMemoryUsage(stats))
          console.log(" > Max mem:", getMaxMemoryUsage(stats))
          console.log(" > GC time:", gcTime.toFixed(2), "ms")
        } else {
          this.forEach(test => console.log(test.toString()))
        }
        console.log("")
        resolve()
      })
      .run({ async: true })
  })
}

module.exports = {
  testCases,
  runTests,
}
