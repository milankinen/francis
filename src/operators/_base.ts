import {
  NOOP_SUBSCRIBER,
  NOOP_SUBSCRIPTION,
  sendEndSafely,
  sendErrorSafely,
  sendEventSafely,
  sendInitialSafely,
  sendNoInitialSafely,
  Source,
  Subscriber,
  Subscription,
} from "../_core"
import { MAX_ORDER } from "../_priority"
import { Transaction } from "../_tx"
import { disableNoUnusedWarning } from "../_util"
import { Scheduler, Task } from "../scheduler/index"

export enum EventType {
  INITIAL = 1,
  NO_INITIAL,
  EVENT,
  ERROR,
  END,
}

/**
 * This function is meant for actual sources (e.g. once, fromArray etc..) to
 * enable multicasting (thus keeping the source's codebase clean)
 *
 * @param source Unicasted "bare" source
 */
export function identity<T>(source: Source<T>): Operator<T, T> {
  return new Identity(source)
}

export abstract class MulticastImplementation<T> implements Subscriber<T> {
  public readonly next: Subscriber<T> = NOOP_SUBSCRIBER

  /**
   * This hook is called every time when operator is activated (gets its first
   * subscriber). If this function returns false, operator aborts the current
   * subscription process. The impelementation may also queue tasks to the given
   * scheduler before returning.
   *
   * @param scheduler
   * @param subscriber
   */
  public beforeActivation(scheduler: Scheduler, subscriber: Subscriber<T>): boolean {
    return true
  }

  /**
   * This hook is called every time when operator is multicasted (gets a new subscriber
   * after initial activation). If this function returns false, operator aborts the current
   * subscription process. The impelementation may also queue tasks to the given
   * scheduler before returning.
   *
   * @param scheduler
   * @param subscriber
   */
  public beforeMulticast(scheduler: Scheduler, subscriber: Subscriber<T>): boolean {
    return true
  }

  public isActive(): boolean {
    return true
  }

  public initial(tx: Transaction, val: T): void {
    this.next.initial(tx, val)
  }

  public noinitial(tx: Transaction): void {
    this.next.noinitial(tx)
  }

  public event(tx: Transaction, val: T): void {
    this.next.event(tx, val)
  }

  public error(tx: Transaction, err: Error): void {
    this.next.error(tx, err)
  }

  public end(tx: Transaction): void {
    this.next.end(tx)
  }
}

const NOOP_MCAST = new class NoopImpl extends MulticastImplementation<any> {}()

export abstract class Operator<A, B>
  implements Subscriber<A>, Source<B>, AbortSubscriptionListener<B> {
  public readonly weight: number

  protected readonly order: number
  protected readonly source: Source<A>
  protected readonly next: MulticastImplementation<B> = NOOP_MCAST

  private __subs: Subscription = NOOP_SUBSCRIPTION
  private __active: boolean = false

  constructor(origin: Source<A>) {
    false && disableNoUnusedWarning(this.__handleDispose, this.__handeReorder)
    this.source = origin
    this.order = MAX_ORDER + 1
    this.weight = origin.weight
  }

  /**
   * Called only internally from Property and EventStream constructors
   * @param impl Multicast implementation to use for this operator
   */
  public setMulticastImplementation(impl: MulticastImplementation<B>): void {
    // tslint:disable-next-line:semicolon whitespace
    ;(this.next as MulticastImplementation<B>) = impl
  }

  /**
   * Called only internally when new EventStream or Property subscribes to this operator
   * during their activation.
   * @param scheduler
   * @param subscriber
   * @param order
   */
  public subscribe(scheduler: Scheduler, subscriber: Subscriber<B>, order: number): Subscription {
    const next = this.next.next
    return next === NOOP_SUBSCRIBER
      ? this.__handleActivation(scheduler, subscriber, order)
      : this.__handleMulticast(scheduler, subscriber, order)
  }

  public abstract initial(tx: Transaction, val: A): void

  public abstract event(tx: Transaction, val: A): void

  public noinitial(tx: Transaction): void {
    this.next.noinitial(tx)
  }

  public error(tx: Transaction, err: Error): void {
    this.next.error(tx, err)
  }

  public end(tx: Transaction): void {
    this.next.end(tx)
  }

  public isActive(): boolean {
    return this.__active
  }

  public handleAbort(subscriber: Subscriber<B>): void {}

  protected activateFirst(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    // TODO
    this.__active = true
    this.__subs = this.source.subscribe(scheduler, this, order)
    return new MulticastSubscription(subscriber, this as any)
  }

  protected activateLate(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    return new MulticastSubscription(subscriber, this as any)
  }

  protected abort(scheduler: Scheduler, subscriber: Subscriber<B>, order: number): Subscription {
    scheduler.scheduleAbortSubscription(new AbortSubscriptionTask(this, subscriber))
    return this.activateLate(scheduler, subscriber, order)
  }

  protected propagateReorder(order: number): void {
    this.__subs.reorder(order)
  }

  protected dispose(): void {
    this.__active = false
    this.__subs.dispose()
    this.__subs = NOOP_SUBSCRIPTION
  }

  protected handleActivation(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    return this.activateFirst(scheduler, subscriber, order)
  }

  protected handleMulticast(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    return this.activateLate(scheduler, subscriber, order)
  }

  protected handleReorder(order: number): void {
    this.propagateReorder(order)
  }

  protected handleDispose(): void {
    this.dispose()
  }

  private __updateNext(next: Subscriber<B>): void {
    // tslint:disable-next-line:semicolon whitespace
    ;(this.next.next as Subscriber<B>) = next
  }

  private __updateOrder(order: number): void {
    // tslint:disable-next-line:semicolon whitespace
    ;(this.order as number) = order
  }

  private __handleActivation(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    if (this.next.beforeActivation(scheduler, subscriber)) {
      this.__updateNext(subscriber)
      this.__updateOrder(order)
      return this.handleActivation(scheduler, subscriber, order)
    } else {
      return NOOP_SUBSCRIPTION
    }
  }

  private __handleMulticast(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    if (this.next.beforeMulticast(scheduler, subscriber)) {
      const next = this.next.next
      if (next instanceof MulticastDelegatee) {
        next.add(subscriber, order)
      } else {
        this.__updateNext(
          new MulticastDelegatee({ s: next, o: this.order }, { s: subscriber, o: order }),
        )
      }
      if (order < this.order) {
        this.__updateOrder(order)
        this.handleReorder(order)
      }
      return this.handleMulticast(scheduler, subscriber, order)
    } else {
      return NOOP_SUBSCRIPTION
    }
  }

  private __handleDispose(subscriber: Subscriber<B>): void {
    const next = this.next.next
    if (next instanceof MulticastDelegatee) {
      const { o: order, s: updated } = next.remove(subscriber)
      this.__updateNext(updated)
      if (order !== this.order) {
        this.__updateOrder(order)
        this.handleReorder(order)
        this.__subs.reorder(order)
      }
    } else {
      this.__updateNext(NOOP_SUBSCRIBER)
      this.__updateOrder(MAX_ORDER + 1)
      this.handleDispose()
    }
  }
  private __handeReorder(subscriber: Subscriber<B>, order: number): void {
    const next = this.next.next
    next instanceof MulticastDelegatee && next.updateOrder(subscriber, order)
    if (order < this.order) {
      this.__updateOrder(order)
      this.handleReorder(order)
    }
  }
}

export interface AbortSubscriptionListener<T> {
  handleAbort(subsciber: Subscriber<T>): void
}

export class AbortSubscriptionTask<T> implements Task {
  constructor(private li: AbortSubscriptionListener<T>, private subscriber: Subscriber<T>) {}

  public run(): void {
    this.li.handleAbort(this.subscriber)
  }
}

const NIL: OrderedSubscriberNode<any> = {
  a: false,
  n: null as any,
  o: MAX_ORDER + 1,
  s: NOOP_SUBSCRIBER,
}

class MulticastDelegatee<T> implements Subscriber<T> {
  private n: number
  private head: OrderedSubscriberNode<T>
  private tail: OrderedSubscriberNode<T>

  constructor(a: OrderedSubscriber<T>, b: OrderedSubscriber<T>) {
    this.head = this.tail = NIL
    this.head = { ...a, a: true, n: this.head }
    this.head = { ...b, a: true, n: this.head }
    this.n = 2
  }

  public add(subscriber: Subscriber<T>, order: number): void {
    this.head = { o: order, s: subscriber, a: true, n: this.head }
    ++this.n
  }

  public remove(subscriber: Subscriber<T>): OrderedSubscriber<T> {
    let prev = NIL
    let next = this.head
    let order = MAX_ORDER + 1
    while (next !== NIL) {
      if (next.s === subscriber) {
        next.a = false
        prev !== NIL ? (prev.n = next.n) : (this.head = next.n)
        --this.n
      } else if (next.o < order) {
        order = next.o
      }
      prev = next
      next = next.n
    }
    return {
      o: order,
      s: this.n > 1 ? this : this.head.s,
    }
  }

  public updateOrder(subscriber: Subscriber<T>, order: number): void {
    let next = this.head
    while (next !== NIL) {
      if (next.s === subscriber) {
        next.o = order
        return
      } else {
        next = next.n
      }
    }
  }

  public isActive(): boolean {
    return true
  }

  public initial(tx: Transaction, val: T): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendInitialSafely(tx, next.s, val)
      next = next.n
    }
  }

  public noinitial(tx: Transaction): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendNoInitialSafely(tx, next.s)
      next = next.n
    }
  }

  public event(tx: Transaction, val: T): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendEventSafely(tx, next.s, val)
      next = next.n
    }
  }

  public error(tx: Transaction, err: Error): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendErrorSafely(tx, next.s, err)
      next = next.n
    }
  }

  public end(tx: Transaction): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendEndSafely(tx, next.s)
      next = next.n
    }
  }
}

interface OrderedSubscriberNode<T> extends OrderedSubscriber<T> {
  n: OrderedSubscriberNode<T>
  a: boolean
}

class MulticastSubscription<T> implements Subscription {
  private s: Subscriber<T>
  private o: MulticastSubscriptionOwner<T>

  // tslint:disable-next-line:ban-types
  constructor(subscriber: Subscriber<T>, owner: MulticastSubscriptionOwner<T>) {
    this.s = subscriber
    this.o = owner
  }

  public dispose(): void {
    this.o.__handleDispose(this.s)
  }

  public reorder(order: number): void {
    this.o.__handeReorder(this.s, order)
  }
}

interface MulticastSubscriptionOwner<T> {
  __handleDispose(subscriber: Subscriber<T>): void
  __handeReorder(subscriber: Subscriber<T>, order: number): void
}

interface OrderedSubscriber<T> {
  s: Subscriber<T>
  o: number
}

class Identity<T> extends Operator<T, T> {
  public initial(tx: Transaction, val: T): void {
    this.next.initial(tx, val)
  }
  public event(tx: Transaction, val: T): void {
    this.next.event(tx, val)
  }
}
