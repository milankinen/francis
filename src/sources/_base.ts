import { __DEVELOPER__, logAndThrow } from "../_assert"
import {
  NOOP_SUBSCRIBER,
  sendRootEnd,
  sendRootError,
  sendRootNext,
  Source,
  Subscriber,
  Subscription,
} from "../_core"
import { AnyEvent } from "../_interfaces"
import { isEnd, isError, isEvent, isNext } from "../Event"
import { scheduleActivationTask, Task } from "../scheduler/index"

export abstract class Root<T> implements Source<T> {
  public readonly weight: number = 1
  public ended: boolean = false
  constructor(public readonly sync: boolean) {}

  public subscribe(subscriber: Subscriber<T>, order: number): Subscription {
    return this.create(subscriber)
  }

  protected abstract create(subscriber: Subscriber<T>): Activation<T, any>
}

export abstract class Activation<T, R extends Root<T>> implements Task, Subscription {
  // the real subscriber
  protected readonly real: Subscriber<T>
  protected readonly owner: R
  protected active: boolean = true
  protected subscriber: Subscriber<T>

  constructor(owner: R, subscriber: Subscriber<T>) {
    this.real = this.subscriber = subscriber
    this.owner = owner
  }

  public run(): void {
    if (this.active) {
      if (this.owner.ended) {
        sendRootEnd(this.subscriber)
      } else {
        if (this.subscriber.begin()) {
          this.start()
        }
      }
    }
  }

  public activate(): void {
    if (this.owner.sync) {
      this.run()
    } else {
      scheduleActivationTask(this)
    }
  }

  public dispose(): void {
    this.active = false
    this.subscriber = NOOP_SUBSCRIBER
    this.stop()
  }

  public reorder(order: number): void {
    // no-op
  }

  protected abstract start(): void

  protected abstract stop(): void

  protected send(event: T | AnyEvent<T>): void {
    if (isEvent(event)) {
      if (isNext(event)) {
        this.sendNext(event.value)
      } else if (isError(event)) {
        this.sendError(event.error)
      } else if (isEnd(event)) {
        this.sendEnd()
      } else {
        if (__DEVELOPER__) {
          logAndThrow("**BUG** event type not known: " + event)
        }
      }
    } else {
      sendRootNext(this.subscriber, event)
    }
  }

  protected sendNext(val: T): void {
    sendRootNext(this.subscriber, val)
  }

  protected sendError(err: Error): void {
    sendRootError(this.subscriber, err)
  }

  protected sendEnd(): void {
    this.owner.ended = true
    sendRootEnd(this.subscriber)
  }
}
