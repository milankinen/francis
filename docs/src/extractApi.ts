import * as fs from "fs"
import * as _ from "lodash"
import { Application, DeclarationReflection, ReflectionKind, SignatureReflection } from "typedoc"
import { Comment, ReferenceType, ReflectionType, Type } from "typedoc/dist/lib/models"

const docs = new Application({ target: "es6", mode: "modules" })
const result = docs.convert(["../src/index.ts"])

interface FrancisParam {
  name: string
  index: number
  description?: string
  from: "signature" | "comment"
}

interface FrancisReturn {
  type: string
  description?: string
}

type FrancisParams = Record<string, FrancisParam>

interface FrancisFunction {
  name: string
  description: string
  signatures: string[]
  params: FrancisParams
  seeAlso: string[]
  category?: string
  returns: FrancisReturn
  example?: string
}

const signatureParams = (sig: SignatureReflection): FrancisParams =>
  ((sig && sig.parameters) || [])
    .map((p, idx) => {
      const comment = p.comment
      return {
        name: p.name,
        index: idx,
        description: comment ? comment.text || comment.shortText : undefined,
        from: "signature",
      }
    })
    .reduce((all, p) => ({ ...all, [p.name]: p }), {})

const commentParams = (comment: Comment): FrancisParams =>
  ((comment && comment.tags) || [])
    .filter(tag => tag.tagName === "param")
    .map((tag, idx) => ({
      name: tag.paramName,
      index: idx,
      description: tag.text,
      from: "comment",
    }))
    .reduce((all, p) => ({ ...all, [p.name]: p }), {})

const selectParam = (p1: FrancisParam, p2: FrancisParam): FrancisParam => {
  const order = (p: FrancisParam) => (p.from === "signature" ? -1000000 : 0) + p.index
  return [p1, p2].sort((a, b) => order(a) - order(b))[0]
}
const combineParams = (...params: FrancisParams[]): FrancisParams =>
  _.chain(params)
    .flatMap(p => _.values(p))
    .reduce((all: FrancisParams, p: FrancisParam) => {
      const prev = p.name in all ? all[p.name] : p
      const next = { ...selectParam(p, prev), description: prev.description || p.description }
      return {
        ...all,
        [p.name]: next,
      }
    }, {})
    .value()

const publicDeclarationsOf = (kind: ReflectionKind) =>
  (result.getReflectionsByKind(kind) as DeclarationReflection[]).filter(
    r => r.flags.isExported && r.flags.isPublic,
  )

const callSignatures = (type: Type): SignatureReflection[] => {
  if (type.type === "reference") {
    const refType = type as ReferenceType
    if (!refType.reflection) {
      return []
    }
    const kind = refType.reflection.kind
    if (kind === ReflectionKind.Interface || kind === ReflectionKind.TypeLiteral) {
      return ((type as ReferenceType).reflection as DeclarationReflection)
        .getAllSignatures()
        .filter(sig => sig.name === "__call")
    }
  } else if (type.type === "reflection") {
    return (type as ReflectionType).declaration
      .getAllSignatures()
      .filter(sig => sig.name === "__call")
  }
  return []
}

const typeToString = (type: Type): string => {
  const curriedCalls = callSignatures(type)
  return curriedCalls.length > 0 ? signatureToString(curriedCalls[0]) : type.toString()
}

const signatureToString = (sig: SignatureReflection): string => {
  const params = (sig.parameters || []).map(p => `${p.name}: ${p.type.toString()}`).join(", ")
  return `(${params}) => ${typeToString(sig.type)}`
}

const firstComment = (sigs: SignatureReflection[]): Comment | undefined =>
  sigs.filter(s => s.comment).map(s => s.comment)[0]

const extractTags = (comment: Comment | undefined, tag: string): string[] =>
  (comment ? comment.tags : [])
    .filter(t => t.tagName === tag && t.text.trim())
    .map(t => t.text.trim())

const functions: FrancisFunction[] = publicDeclarationsOf(ReflectionKind.Function).map(
  reflection => {
    const sigs = reflection
      .getAllSignatures()
      .sort((a, b) => (a.parameters || []).length - (b.parameters || []).length)
      .reverse()
    const comment = reflection.comment || firstComment(sigs)
    return {
      name: reflection.name,
      description: comment ? comment.text || comment.shortText : "",
      signatures: sigs.map(signatureToString),
      params: combineParams(commentParams(reflection.comment), ...sigs.map(signatureParams)),
      category: extractTags(comment, "category")[0],
      seeAlso: extractTags(comment, "see"),
      returns: {
        type: typeToString(sigs[0].type),
        description: extractTags(comment, "returns")[0],
      },
      example: extractTags(comment, "example")[0],
    }
  },
)

const curriedFunctions: FrancisFunction[] = publicDeclarationsOf(ReflectionKind.Variable)
  .filter(reflection => callSignatures(reflection.type).length > 0)
  .map(reflection => {
    const sigs = callSignatures(reflection.type)
      .sort((a, b) => (a.parameters || []).length - (b.parameters || []).length)
      .reverse()
    const comment = reflection.comment || firstComment(sigs)
    return {
      name: reflection.name,
      description: comment ? comment.shortText : "",
      signatures: sigs.map(signatureToString),
      params: combineParams(commentParams(reflection.comment), ...sigs.map(signatureParams)),
      category: extractTags(comment, "category")[0],
      seeAlso: extractTags(comment, "see"),
      returns: {
        type: typeToString(sigs[0].type),
        description: extractTags(comment, "returns")[0],
      },
      example: extractTags(comment, "example")[0],
    }
  })

const data = {
  functions: [...functions, ...curriedFunctions],
}

fs.writeFileSync("api.json", JSON.stringify(data, null, 2))
