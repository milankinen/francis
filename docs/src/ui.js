import * as _ from "lodash"
import React from "react"
import Markdown from "react-markdown/with-html"
import SyntaxHighlighter, { registerLanguage } from "react-syntax-highlighter/dist/prism-light"
import tsLang from "react-syntax-highlighter/dist/languages/prism/typescript"
import baseCodeStyle from "react-syntax-highlighter/dist/styles/prism/base16-ateliersulphurpool.light"

registerLanguage("ts", tsLang)

const codeStyle = _.extend(
  _.transform(baseCodeStyle, (res, value, key) =>
    _.extend(res, { [key]: _.omit(value, "background") }),
  ),
  {
    'pre[class*="language-"]': {
      padding: "15px",
      borderRadius: "5px",
      background: "#eef2f7",
      fontSize: "13px",
      overflowX: "auto",
    },
  },
)
codeStyle['code[class*="language-"]'].fontSize = "13px"

const CodeBlock = ({ code, language = "ts" }) => (
  <SyntaxHighlighter language={language} style={codeStyle}>
    {code}
  </SyntaxHighlighter>
)

const InlineCode = ({ children }) => <code className="siimple-code">{children}</code>

const Md = ({ value }) => (
  <Markdown
    source={value}
    escapeHtml={false}
    renderers={{
      code: ({ children }) => <CodeBlock code={children} />,
      inlineCode: ({ children }) => <InlineCode>{children}</InlineCode>,
    }}
  />
)

const Navbar = ({ version }) => (
  <div id="-top-navi" className="siimple-navbar siimple-navbar--dark siimple-navbar--fluid">
    <button id="sidebar-toggle" className="siimple-btn siimple-btn--navy">
      <i className="fas fa-bars" />
    </button>
    <a className="siimple-navbar-title" href=".">
      Francis
    </a>
    <span className="siimple-navbar-subtitle">{version}</span>
    <div className="siimple--float-right">
      <a
        id="github-btn"
        className="siimple-navbar-item"
        href="https://github.com/milankinen/francis"
      >
        <i className="fab fa-github fa-lg" />
      </a>
    </div>
  </div>
)

const SidebarLink = ({ href, title }) => (
  <a href={`#${href}`} className="link siimple-small siimple--text-bold">
    {title}
  </a>
)

const SearchableGroup = ({ title, items }) => (
  <article className="searchable">
    <h4>{title}</h4>
    {items}
    <span
      className="no-matches siimple-small siimple--text-bold"
      style={{ color: "#546778", display: "none" }}
    >
      No matches
    </span>
  </article>
)

const Sidebar = ({ functions }) => (
  <section id="-side-navi" className="siimple--bg-light">
    <input id="-search" type="text" className="siimple-input" placeholder="Search..." />
    {/* <SearchableGroup title="Classes" items={[]} /> */}
    <SearchableGroup
      title="Functions"
      items={_.chain(functions)
        .map(({ name }, i) => <SidebarLink key={i} href={name.toLowerCase()} title={name} />)
        .value()}
    />
    {/* <SearchableGroup title="Interfaces" items={[]} /> */}
  </section>
)

const FunctionDoc = ({ name, description, signatures, params, returns, seeAlso, example }) => (
  <div className="doc siimple-card" id={name.toLowerCase()}>
    <div className="siimple-card-body">
      <div className="siimple-card-title">{name}</div>
      <CodeBlock code={signatures.join("\n")} />
      <Md value={description} />
      <h4>Params</h4>
      {_.keys(params).length > 0 ? (
        <table className="params">
          <tbody>
            {_.chain(params)
              .values()
              .sortBy(p => p.index)
              .map(({ name, description }, i) => (
                <tr key={i} className="siimple-small">
                  <td className="name siimple--text-bold">{name}</td>
                  <td>
                    <Md value={description} />
                  </td>
                </tr>
              ))
              .value()}
          </tbody>
        </table>
      ) : (
        "None"
      )}
      <h4>
        Returns <InlineCode>{returns.type}</InlineCode>
      </h4>
      <div className="returns siimple-small">
        {returns.description ? <Md value={returns.description} /> : null}
      </div>
      {seeAlso.length ? <h4>See also</h4> : null}
      {seeAlso.map((name, i) => (
        <a key={i} href={`#${name.toLowerCase()}`} className="siimple-small">
          {name}
        </a>
      ))}
      {example ? <h4>Example</h4> : null}
      {example ? <CodeBlock code={example} /> : null}
    </div>
  </div>
)

const Main = ({ functions }) => (
  <section id="-main">
    {functions.map((f, i) => (
      <FunctionDoc key={i} {...f} />
    ))}
  </section>
)

export const UI = ({ version, api: { functions } }) => {
  const sortedFunctions = _.chain(functions)
    .sortBy(f => f.name)
    .value()
  return (
    <div id="-app" className="navi-open">
      <Navbar version={version} />
      <div id="-content">
        <Main functions={sortedFunctions} />
        <Sidebar functions={sortedFunctions} />
      </div>
    </div>
  )
}

/**
 * We want to have real static html site without extra JS libraries so we're gonna
 * do the interactions (toggle and search) by using vanilla JS and inject this script
 * to the rendered page source.
 */
export function setupInteractions() {
  var app = document.getElementById("-app")
  var toggle = document.getElementById("sidebar-toggle")
  var search = document.getElementById("-search")
  toggle.addEventListener("click", function() {
    if (app.classList.contains("navi-open")) {
      app.classList.remove("navi-open")
    } else {
      app.classList.add("navi-open")
    }
  })
  search.addEventListener("focus", function() {
    search.setSelectionRange(0, search.value.length)
  })
  search.addEventListener("input", function(event) {
    var query = (event.target.value || "").trim().toLowerCase()
    document.querySelectorAll("#-side-navi article.searchable").forEach(function(group) {
      var hasMatches = false
      group.querySelectorAll(".link").forEach(function(link) {
        if (query.length === 0 || link.textContent.toLowerCase().indexOf(query) !== -1) {
          link.style.display = "block"
          hasMatches = true
        } else {
          link.style.display = "none"
        }
      })
      var noMatch = group.querySelector(".no-matches")
      noMatch.style.display = hasMatches ? "none" : "block"
    })
  })
}
