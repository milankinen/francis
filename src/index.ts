// interfaces and classes
export * from "./Event"
export * from "./_interfaces"
export { Observable } from "./Observable"
export { EventStream } from "./EventStream"
export { AnyObs, Property } from "./Property"

// factory functions
export { once } from "./sources/once"
export { constant } from "./sources/constant"
export { fromArray } from "./sources/fromArray"

// operators
export { map } from "./operators/map"
export { filter } from "./operators/filter"
export { take } from "./operators/take"
export { first } from "./operators/first"
export { zipAsArray } from "./operators/zip"
export { combineAsArray } from "./operators/combine"

// subscribers / side-effects
export { subscribe } from "./operators/subscribe"

export { log } from "./operators/log"
