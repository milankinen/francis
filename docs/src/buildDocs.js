import * as fs from "fs"
import React from "react"
import { renderToStaticMarkup } from "react-dom/server"
import { UI, setupInteractions } from "./ui"

const api = require("../api.json")
const version = require("../../package.json").version

const intro = fs.readFileSync("./INTRODUCTION.md").toString()

const setupInteractionsOnPageLoad = () =>
  `document.addEventListener("DOMContentLoaded", ${setupInteractions.toString()});`

const script = code => <script type="text/javascript" dangerouslySetInnerHTML={{ __html: code }} />

const APIDocs = () => (
  <html>
    <head>
      <title>Francis API docs</title>
      <link
        rel="stylesheet"
        href="https://cdn.jsdelivr.net/npm/siimple@3.2.0/dist/siimple.min.css"
      />
      <link
        rel="stylesheet"
        href="https://use.fontawesome.com/releases/v5.6.3/css/all.css"
        integrity="sha384-UHRtZLI+pbxtHCWp1t77Bi1L4ZtiqrqD80Kn4Z8NTSRyMA2Fd33n5dQ8lWUE00s/"
        crossOrigin="anonymous"
      />
      <link rel="stylesheet" href="./app.css" />
    </head>
    <body>
      <UI version={version} api={api} intro={intro} />
      {script(setupInteractionsOnPageLoad())}
    </body>
  </html>
)

fs.writeFileSync("index.html", renderToStaticMarkup(<APIDocs />))
