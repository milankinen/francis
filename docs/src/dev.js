import React from "react"
import { render } from "react-dom"
import { UI, setupInteractions } from "./ui"

const api = require("../api.json")
const intro = fetch(require("../INTRODUCTION.md")).then(x => x.text())

class DevUI extends React.Component {
  constructor(props) {
    super(props)
    this.state = {}
  }
  componentDidMount() {
    intro.then(i => this.setState({ intro: i }))
    setTimeout(setupInteractions, 0)
  }
  render() {
    const { intro } = this.state
    return intro ? <UI api={api} intro={intro} version="DEV" /> : null
  }
}

render(<DevUI />, document.getElementById("dev-app"))
