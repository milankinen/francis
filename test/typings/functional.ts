import * as F from "../../src/index"
import { run } from "../_base"

describe("Francis functional interface", () => {
  it("supports curried ops and piping the curried ops", () => {
    const recording = run(record => {
      const recordStr = F.subscribe<string>(record)
      const recordStrObs = (obs: F.Observable<string>) => recordStr(obs)
      const every10ms = F.interval(10)
      const result = F.pipe(
        every10ms("!"),
        F.scan("TSERS", (sum, n) => sum + n),
        F.skip(1),
        F.take(5),
        F.map(s => s.toLowerCase()),
      )
      recordStrObs(result)
    })
    // actually the result is not so important, more important is that TS compiler won't break
    expect(recording).toMatchSnapshot()
  })
})
