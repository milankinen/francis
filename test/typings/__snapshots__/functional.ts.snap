// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Francis functional interface supports curried ops and piping the curried ops 1`] = `
Array [
  Next {
    "value": "tsers!",
  },
  Next {
    "value": "tsers!!",
  },
  Next {
    "value": "tsers!!!",
  },
  Next {
    "value": "tsers!!!!",
  },
  Next {
    "value": "tsers!!!!!",
  },
  End {},
]
`;
